File: problems_claude/12.py
# # Problem 12: Container With Most Water
# **Difficulty: ** Medium
# **Time: ** 20 minutes
# **Pattern: ** Two pointers(optimization problem)
#
# Given an array of heights, find two lines that together with the x - axis
# form a container that holds the most water.


def max_water_container(heights):
    """
    Find maximum water that can be contained.

    Example:
    Input: [1, 8, 6, 2, 5, 4, 8, 3, 7]
    Output: 49
    Explanation: Lines at index 1 (height=8) and index 8 (height=7)
                 form container with area = 7 * 7 = 49

    The area is calculated as:
    min(height[i], height[j]) * (j - i)
    """
    # Your code here
    pass


# Test cases
assert max_water_container([1, 8, 6, 2, 5, 4, 8, 3, 7]) == 49
assert max_water_container([1, 1]) == 1
assert max_water_container([4, 3, 2, 1, 4]) == 16
assert max_water_container([1, 2, 1]) == 2
print("All tests passed!")

File: cpp/problems_2/12.cpp
// #### Problem 12: Reverse String
//
// **Difficulty:** Easy
// **Time:** 10 minutes
// **Pattern:** Two pointers
//
// Write a function that reverses a string. The input string is given as an
// array of characters `s`. You must do this by modifying the input array
// in-place with O(1) extra memory.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

void reverseString(std::vector<char> &s) {
  /*
  Reverse string in-place.

  Examples:
  Input: s = ["h","e","l","l","o"]
  Output: ["o","l","l","e","h"]

  Input: s = ["H","a","n","n","a","h"]
  Output: ["h","a","n","n","a","H"]
  */
  // Your code here
  auto left_it = s.begin();
  auto right_it = s.end() - 1;

  while (left_it < right_it) {
    std::iter_swap(left_it, right_it);

    left_it++;
    right_it--;
  }
}

int main() {
  std::vector<char> s1 = {'h', 'e', 'l', 'l', 'o'};
  reverseString(s1);
  assert(s1 == std::vector<char>({'o', 'l', 'l', 'e', 'h'}));

  std::vector<char> s2 = {'H', 'a', 'n', 'n', 'a', 'h'};
  reverseString(s2);
  assert(s2 == std::vector<char>({'h', 'a', 'n', 'n', 'a', 'H'}));

  std::cout << "Problem 12: All tests passed!" << std::endl;
  return 0;
}
